### ISSUE TITLE
[java] Junit rule to enforce using messages for assertTrue/False, etc.

### ISSUE THREAD

**Issue by ARentz07** [pmd/pmd#366]
**Rule Set:** Maybe junit, though I don't know everyone would want this.

**Description:** This issue proposes adding a new rule to add messages to JUnit's assertion methods in tests. The motivation is that having an `assertTrue` call with a complex condition inside provides no insight to a developer in the event of a failure, and we could improve this situation by requiring a message to be used.

This could also, as a side effect, encourage use of things like `assertThat`, which automatically provides these types of messages. However, it's not always appropriate to do so. In those cases when it isn't, we should provide the most helpful message possible. 

**Code Sample demonstrating the issue:** Not really applicable here, but for instance, a somewhat contrived example:

```java
assertTrue(haystack.contains("needle"));
```

Would be required to be updated to:

```java
assertTrue(haystack.contains("needle"), "Expected to find needle in haystack, but haystack was " + haystack);
```
... or something similar.

**Running PMD through:** [CLI | Ant | __**Maven**__ | Gradle | Designer | Other]

**Comment by jsotuyod** [comment]
This should already be covered by [java-junit/JUnitAssertionsShouldIncludeMessage](https://pmd.github.io/pmd-5.5.6/pmd-java/rules/java/junit.html#JUnitAssertionsShouldIncludeMessage). Is there any reason that rule wouldn't work for you?

**Comment by ARentz07** [comment]
Mind blown. We have been using PMD for quite some time on my team; I don't know why we aren't using this. I'll try enabling this and verify.

**Comment by jsotuyod** [comment]
@ARentz07 ok, let me know if we should close this or reword it.

**Comment by ARentz07** [comment]
We can close this. There were several layers to why this wasn't working through the maven-pmd-plugin:
1) The `includeTests` config is only valid for the `pmd` and `cpd` goals, not the `check` or `cpd-check` goals. I had to specify both in my parent pom for the property to be recognized.
2) There seems to be an issue with the plugin in which violations in the test code will not cause the build to fail. The violations still get written to the pmd.xml file, but the build is successful.

I'm still trying to work through these issues with my maven project, but they are unrelated to the rules.
